### PR TITLE
OutOfMemoryError Test Case: Where 80,000,000 coming from

### PR DESCRIPTION
**Pull request by suztomo**
```
mvn test -Dtest=DependencyGraphBuilderTest
```

### CODE REVIEW

**Comment by suztomo** [comment]
My MacBook complained:

```
org.eclipse.aether.resolution.DependencyResolutionException: The following artifacts could not be resolved: com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, xerces:xerces-impl:jar:2.6.2, xml-apis:xml-apis:jar:2.6.2, com.jcraft:jsch.agentproxy:jar:0.0.6, maven-plugins:maven-cobertura-plugin:plugin:1.3, maven-plugins:maven-findbugs-plugin:plugin:1.3.1, xerces:xerces-impl:jar:2.6.2, xml-apis:xml-apis:jar:2.6.2, com.googlecode:openbeans:jar:1.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, xerces:xerces-impl:jar:2.6.2, xml-apis:xml-apis:jar:2.6.2, xerces:xerces-impl:jar:2.6.2, xml-apis:xml-apis:jar:2.6.2, xerces:xerces-impl:jar:2.6.2, xml-apis:xml-apis:jar:2.6.2, com.jcraft:jsch.agentproxy:jar:0.0.6, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, javax.sql:jdbc-stdext:jar:2.0, javax.sql:jdbc-stdext:jar:2.0, xerces:xerces-impl:jar:2.6.2, xml-apis:xml-apis:jar:2.6.2, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, xerces:xerces-impl:jar:2.6.2, xml-apis:xml-apis:jar:2.6.2, xerces:xerces-impl:jar:2.6.2, xml-apis:xml-apis:jar:2.6.2, xerces:xerces-impl:jar:2.6.2, xml-apis:xml-apis:jar:2.6.2, xerces:xerces-impl:jar:2.6.2, xml-apis:xml-apis:jar:2.6.2, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, xerces:xerces-impl:jar:2.6.2, xml-apis:xml-apis:jar:2.6.2: Could not find artifact com.google.inject:guice:jar:no_deps:3.0 in  (https://repo1.maven.org/maven2/)

	at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveDependencies(DefaultRepositorySystem.java:357)
	at com.google.cloud.tools.opensource.dependencies.DependencyGraphBuilder.resolveCompileTimeDependencies(DependencyGraphBuilder.java:151)
	at com.google.cloud.tools.opensource.dependencies.DependencyGraphBuilderTest.testResolveCompileTimeDependenciesForBeamHCatalog(DependencyGraphBuilderTest.java:319)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:230)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:58)
Caused by: org.eclipse.aether.resolution.ArtifactResolutionException: The following artifacts could not be resolved: com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, xerces:xerces-impl:jar:2.6.2, xml-apis:xml-apis:jar:2.6.2, com.jcraft:jsch.agentproxy:jar:0.0.6, maven-plugins:maven-cobertura-plugin:plugin:1.3, maven-plugins:maven-findbugs-plugin:plugin:1.3.1, xerces:xerces-impl:jar:2.6.2, xml-apis:xml-apis:jar:2.6.2, com.googlecode:openbeans:jar:1.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, xerces:xerces-impl:jar:2.6.2, xml-apis:xml-apis:jar:2.6.2, xerces:xerces-impl:jar:2.6.2, xml-apis:xml-apis:jar:2.6.2, xerces:xerces-impl:jar:2.6.2, xml-apis:xml-apis:jar:2.6.2, com.jcraft:jsch.agentproxy:jar:0.0.6, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, javax.sql:jdbc-stdext:jar:2.0, javax.sql:jdbc-stdext:jar:2.0, xerces:xerces-impl:jar:2.6.2, xml-apis:xml-apis:jar:2.6.2, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, xerces:xerces-impl:jar:2.6.2, xml-apis:xml-apis:jar:2.6.2, xerces:xerces-impl:jar:2.6.2, xml-apis:xml-apis:jar:2.6.2, xerces:xerces-impl:jar:2.6.2, xml-apis:xml-apis:jar:2.6.2, xerces:xerces-impl:jar:2.6.2, xml-apis:xml-apis:jar:2.6.2, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, com.google.inject:guice:jar:no_deps:3.0, xerces:xerces-impl:jar:2.6.2, xml-apis:xml-apis:jar:2.6.2: Could not find artifact com.google.inject:guice:jar:no_deps:3.0 in  (https://repo1.maven.org/maven2/)
	at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve(DefaultArtifactResolver.java:424)
	at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifacts(DefaultArtifactResolver.java:229)
	at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveDependencies(DefaultRepositorySystem.java:340)
	... 27 more
Caused by: org.eclipse.aether.transfer.ArtifactNotFoundException: Could not find artifact com.google.inject:guice:jar:no_deps:3.0 in  (https://repo1.maven.org/maven2/)
	at org.eclipse.aether.connector.basic.ArtifactTransportListener.transferFailed(ArtifactTransportListener.java:48)
	at org.eclipse.aether.connector.basic.BasicRepositoryConnector$TaskRunner.run(BasicRepositoryConnector.java:369)
	at org.eclipse.aether.util.concurrency.RunnableErrorForwarder$1.run(RunnableErrorForwarder.java:75)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)


Process finished with exit code 255

```

**Comment by suztomo** [comment]
It seems the tree representation is not the only problem for beam hcatalog. The 79,447 artifacts resulted in 570 JAR files for the class path for Linkage Checker, which it cannot handle.

```
mvn test -Dtest=LinkageCheckerTest#testBeamHCatalogArtifact
```

This fails with OutOfMemoryError when reading 570 JAR files. Stacktrace:

```
java.lang.OutOfMemoryError: GC overhead limit exceeded

	at com.google.cloud.tools.opensource.classpath.ClassPathEntry.hashCode(ClassPathEntry.java:80)
	at java.util.Arrays.hashCode(Arrays.java:4146)
	at java.util.Objects.hash(Objects.java:128)
	at com.google.cloud.tools.opensource.classpath.ClassFile.hashCode(ClassFile.java:78)
	at java.util.HashMap.hash(HashMap.java:339)
	at java.util.HashMap.put(HashMap.java:612)
	at com.google.common.collect.ImmutableMultimap$Builder.put(ImmutableMultimap.java:165)
	at com.google.common.collect.ImmutableSetMultimap$Builder.put(ImmutableSetMultimap.java:262)
	at com.google.cloud.tools.opensource.classpath.SymbolReferenceMaps$Builder.addClassReference(SymbolReferenceMaps.java:68)
	at com.google.cloud.tools.opensource.classpath.ClassDumper.findSymbolReferences(ClassDumper.java:204)
	at com.google.cloud.tools.opensource.classpath.ClassDumper.findSymbolReferences(ClassDumper.java:166)
	at com.google.cloud.tools.opensource.classpath.LinkageChecker.create(LinkageChecker.java:86)
	at com.google.cloud.tools.opensource.classpath.LinkageChecker.create(LinkageChecker.java:69)
	at com.google.cloud.tools.opensource.classpath.LinkageCheckerTest.testBeamHCatalogArtifact(LinkageCheckerTest.java:1131)


```

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java [hunk 1]
```diff
@@ -146,8 +150,12 @@ boolean isSystemClass(String className) {
   SymbolReferenceMaps findSymbolReferences() throws IOException {
     SymbolReferenceMaps.Builder builder = new SymbolReferenceMaps.Builder();
 
+    int count = 0;
     for (ClassPathEntry jar : inputClassPath) {
-      for (JavaClass javaClass : listClasses(jar)) {
+      count++;
+      ImmutableSet<JavaClass> javaClasses = listClasses(jar);
+      System.out.println(count + " : " + jar + ", with " + javaClasses.size() + " classes.");
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java:157]: For the OutOfMemoryError while reading 570 JAR files, the last output was presto's hadoop.

```
1 : org.apache.beam:beam-sdks-java-io-hcatalog:2.20.0, with 26 classes.
2 : org.apache.beam:beam-vendor-guava-26_0-jre:0.1, with 1951 classes.
3 : org.apache.beam:beam-sdks-java-core:2.20.0, with 3480 classes.
4 : org.apache.beam:beam-sdks-java-io-hadoop-common:2.20.0, with 7 classes.
5 : org.slf4j:slf4j-api:1.7.25, with 34 classes.
6 : com.fasterxml.jackson.core:jackson-annotations:2.10.2, with 69 classes.
7 : com.fasterxml.jackson.core:jackson-core:2.10.2, with 118 classes.
8 : com.fasterxml.jackson.core:jackson-databind:2.10.2, with 656 classes.
9 : org.apache.hadoop:hadoop-common:2.8.5, with 2263 classes.
10 : org.apache.hive:hive-exec:2.1.0, with 17491 classes.
11 : org.apache.hive.hcatalog:hive-hcatalog-core:2.1.0, with 115 classes.
12 : org.apache.beam:beam-model-pipeline:2.20.0, with 607 classes.
13 : org.apache.beam:beam-model-job-management:2.20.0, with 220 classes.
14 : org.apache.beam:beam-vendor-bytebuddy-1_10_8:0.1, with 2359 classes.
15 : com.google.code.findbugs:jsr305:3.0.2, with 35 classes.
16 : org.apache.avro:avro:1.8.2, with 1172 classes.
17 : org.xerial.snappy:snappy-java:1.1.4, with 28 classes.
18 : joda-time:joda-time:2.10.5, with 247 classes.
19 : org.tukaani:xz:1.8, with 116 classes.
20 : junit:junit:4.13-beta-3, with 347 classes.
21 : org.hamcrest:hamcrest-core:2.1, with 1 classes.
22 : org.hamcrest:hamcrest-library:2.1, with 1 classes.
23 : io.airlift:aircompressor:0.16, with 94 classes.
24 : com.facebook.presto.hadoop:hadoop-apache2:3.2.0-1, with 17869 classes.

java.lang.OutOfMemoryError: GC overhead limit exceeded

```

Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java:157]: The reason why hamcrest-core only has 1 class:
```
suztomo-macbookpro44% jar tf /Users/suztomo/.m2/repository/org/hamcrest/hamcrest-core/2.1/hamcrest-core-2.1.jar 
META-INF/
META-INF/MANIFEST.MF
org/
org/hamcrest/
org/hamcrest/HamcrestCoreIsDeprecated.class
README.txt
```

**Comment by elharo** [comment]
It feels like we should be able to handle 570 jars unless we have quadratic algorithms somewhere. (We might: we're testing everything against everything.) 

How bad is the OOME? E.g do we need a plausibly larger heap size or a ridiculously larger heap size?

**Comment by suztomo** [comment]
It passed with 4GB heap size. Not that bad.
It got stuck with 2 GB heap.

File: dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilderTest.java [hunk 2]
```diff
@@ -305,4 +316,102 @@ public void testDependencyPathRoot_twoDependency() {
...
+                  new DefaultArtifact("org.apache.beam:beam-sdks-java-io-hcatalog:2.20.0"))),
+          true);
+      fail();
+    } catch (DependencyResolutionException ex) {
+      DependencyResult result = ex.getResult();
+      root = result.getRoot();
+    }
+
+    int countByGraphTraverser = countByTraverserForGraph(root);
+    System.out.println("Count by Traverser.forGraph: " + countByGraphTraverser);
```
Comment (elharo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilderTest.java:341]: can we get rid of these printlns now?

File: dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilderTest.java [hunk 3]
```diff
@@ -305,4 +316,102 @@ public void testDependencyPathRoot_twoDependency() {
...
+  int countByBfs(DependencyNode root) {
+    // Count the number of nodes
+    ArrayDeque<DependencyNode> queue = new ArrayDeque<>();
+    queue.add(root);
+    int count = 0;
+
+    while (!queue.isEmpty()) {
+      DependencyNode node = queue.poll();
+      count++;
+      if (count % 1_000_000 == 0) {
```
Comment (elharo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilderTest.java:408]: empty block

**Comment by suztomo** [comment]
@elharo I'm not going to merge this PR to master.

**Comment by suztomo** [comment]
Trying to run original problematic test case with bigger heap.
https://github.com/GoogleCloudPlatform/cloud-opensource-java/commit/99852024d32a5a19f0fedeff630fd5ab1f04d710

**Comment by suztomo** [comment]
I waited 12 hours but the test case doesn't return https://github.com/GoogleCloudPlatform/cloud-opensource-java/commit/99852024d32a5a19f0fedeff630fd5ab1f04d710 . Now I'll see where it gets stuck.

**Comment by suztomo** [comment]
CircleCI build now requires authentication with GitHub account. Strange.

**Comment by suztomo** [comment]
Trying 12 GB for the test case https://github.com/GoogleCloudPlatform/cloud-opensource-java/commit/99852024d32a5a19f0fedeff630fd5ab1f04d710 . Started 11:18.

**Pull request closed without merging**